### PR TITLE
trie: improve range proof

### PR DESCRIPTION
**Pull request by rjl493456442**
In this PR, range proof is improved.

Originally we only allow range proof with two existent edge proofs. 
However in the real use cases, it's quite common that the first edge
proof might point to a non-existent node. So in this PR, non-existent
left edge proof is allowed.

Besides, there are a few corner cases handled:
- The range is empty, and the left edge proof is a non-existent proof. In this case 
   the function will verify whether there still exist some leaves on the right side. If
   so, then return the error.
- The range only has one element, no matter the left edge proof is existent one or not
- The range has all the elements, the edge proofs can be nil

### CODE REVIEW

File: trie/proof.go [hunk 1]
```diff
@@ -241,50 +260,165 @@ func unsetInternal(node node, left []byte, right []byte) error {
...
+//   case if there are still some other leaves available on the right side, then
+//   an error will be returned.
+//
+// - One element proof. In this case no matter the left edge proof is a non-existent
+//   proof or not, we can always verify the correctness of the proof.
+func VerifyRangeProof(rootHash common.Hash, firstKey []byte, keys [][]byte, values [][]byte, firstProof ethdb.KeyValueReader, lastProof ethdb.KeyValueReader) error {
 	if len(keys) != len(values) {
 		return fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values))
 	}
+	// Speical case, there is no edge proof at all. Then the
```
Comment (holiman) [inline trie/proof.go:367]:
```suggestion
	// Special case, there is no edge proof at all. Then the
```

File: trie/proof.go [hunk 2]
```diff
@@ -241,50 +260,165 @@ func unsetInternal(node node, left []byte, right []byte) error {
...
+	if firstProof == nil && lastProof == nil {
+		emptytrie, err := New(common.Hash{}, NewDatabase(memorydb.New()))
+		if err != nil {
+			return err
+		}
+		for index, key := range keys {
+			emptytrie.TryUpdate(key, values[index])
+		}
+		if emptytrie.Hash() != rootHash {
+			return fmt.Errorf("invalid proof, wanthash %x, got %x", rootHash, emptytrie.Hash())
```
Comment (holiman) [inline trie/proof.go:378]:
```suggestion
			return fmt.Errorf("invalid proof, want hash %x, got %x", rootHash, emptytrie.Hash())
```

File: trie/proof.go [hunk 3]
```diff
@@ -241,50 +260,165 @@ func unsetInternal(node node, left []byte, right []byte) error {
...
+		}
+		return nil
+	}
+	// Special case, there is a provided non-existent proof and
+	// zero leaf, it means no more leaf we can get in the trie.
```
Comment (holiman) [inline trie/proof.go:383]: I don't understand the sentence "it means no more leaf we can get in the trie". Can you please reformulate or clarify that?

**Comment by holiman** [comment]
Playing with the prover a bit, I found this a bit odd.. In the testcase below, if I just mutate one of the values, the verification succeeds (the test fails). But if I mutate it "a lot", then the verification fails (test passes):

```diff
diff --git a/trie/proof_test.go b/trie/proof_test.go
index 2562f4f8b8..3db20ffc8d 100644
--- a/trie/proof_test.go
+++ b/trie/proof_test.go
@@ -118,7 +118,34 @@ func TestBadProof(t *testing.T) {
 
 			mutateByte(val)
 			proof.Put(crypto.Keccak256(val), val)
+			if _, err := VerifyProof(root, kv.k, proof); err == nil {
+				t.Fatalf("prover %d: expected proof to fail for key %x", i, kv.k)
+			}
+		}
+	}
+}
 
+func TestBadProofBadValue(t *testing.T) {
+	trie, vals := randomTrie(800)
+	root := trie.Hash()
+	for i, prover := range makeProvers(trie) {
+		for _, kv := range vals {
+			proof := prover(kv.k)
+			if proof == nil {
+				t.Fatalf("prover %d: nil proof", i)
+			}
+			it := proof.NewIterator(nil, nil)
+			for i, d := 0, mrand.Intn(proof.Len()); i <= d; i++ {
+				it.Next()
+			}
+			key := it.Key()
+			val, _ := proof.Get(key)
+			it.Release()
+			mutateByte(val)
+			// This is correctly detected:
+			// proof.Put(key, []byte("donald duck donald duck donald duck donald duck donald duck"))
+			// But not this:
+			proof.Put(key,val)
 			if _, err := VerifyProof(root, kv.k, proof); err == nil {
 				t.Fatalf("prover %d: expected proof to fail for key %x", i, kv.k)
 			}
```

**Comment by holiman** [comment]
This makes my test pass (verification fail): 
```diff
diff --git a/trie/proof.go b/trie/proof.go
index d02596b7f2..f818550059 100644
--- a/trie/proof.go
+++ b/trie/proof.go
@@ -115,6 +115,16 @@ func VerifyProof(rootHash common.Hash, key []byte, proofDb ethdb.KeyValueReader)
                if err != nil {
                        return nil, fmt.Errorf("bad proof node %d: %v", i, err)
                }
+               {// Check hash is correct
+                       hasher := newHasher(false)
+                       defer returnHasherToPool(hasher)
+                       hash := hasher.hashData(buf)
+                       if !bytes.Equal(hash, wantHash[:]){
+                               return nil, fmt.Errorf("proof node %d wrong has, got %x want %x", i, hash, wantHash)
+                       }
+               }
+
+
                keyrest, cld := get(n, key, true)
                switch cld := cld.(type) {
                case nil:
```

File: trie/proof.go [hunk 4]
```diff
@@ -241,50 +260,165 @@ func unsetInternal(node node, left []byte, right []byte) error {
...
+			// Find the fork point, it's an non-existent branch.
+			if removeLeft {
+				return errors.New("invalid right edge proof")
+			}
+			if bytes.Compare(cld.Key, key[pos:]) > 0 {
+				// The key of fork shortnode is greater than the
+				// path(it belongs to the range), unset the entrie
+				// branch. The parent must be a fullnode.
+				fn := parent.(*fullNode)
+				fn.Children[key[pos-1]] = nil
```
Comment (karalabe) [inline trie/proof.go:312]: Shouldn't we set `fn` to firty here?

File: trie/proof.go [hunk 5]
```diff
@@ -241,50 +260,165 @@ func unsetInternal(node node, left []byte, right []byte) error {
...
+			} else {
+				// The key of fork shortnode is less than the
+				// path(it doesn't belong to the range), keep
+				// it with the cached hash available.
+				return nil
+			}
+		}
+		if _, ok := cld.Val.(valueNode); ok {
+			fn := parent.(*fullNode)
+			fn.Children[key[pos-1]] = nil
```
Comment (karalabe) [inline trie/proof.go:322]: Shouldn't we set `fn` to firty here?

File: trie/proof.go (hunk 4 above)
Comment (rjl493456442) [reply trie/proof.go:312]: because the fn here refers to the parent, it's already marked as dirty

File: trie/proof.go (hunk 5 above)
Comment (rjl493456442) [reply trie/proof.go:322]: because the fn here refers to the parent, it's already marked as dirty

**Pull request merged**